### PR TITLE
Embeddings and frontend

### PR DESCRIPTION
**Pull request by earny-joe**
The following pull request includes the following updates:

- addition of Python script that performed tests on embedding models
- relocation of streamlit visualization script to `src` folder
- initial development of v2 of streamlit within `src` folder
- updates to requirements.txt to reflect new Python libraries to be used
- relocation of the script that converts raw JSON to BigQuery ready parquet files to `src/data` subfolder

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -191,11 +191,12 @@ rsa==4.0
 s3transfer==0.3.3
```
Comment (rkingery) [inline requirements.txt:191]: Note your `requirements.txt` should contain the *minimal* libraries your *users* need to have installed to run your app, and that you need to regenerate data and models. It shouldn't contain any libraries that you used for experiments but never use in the source code itself. I recommend either culling this list down to the essentials for source code work, or renaming this file to `requirements_experiments.txt` and creating a new `requirements.txt` that only contains what you need for the source code to run.

File: src/visualization/streamlit-exp-v2.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+'''
```
Comment (rkingery) [inline src/visualization/streamlit-exp-v2.py:1]: Why do you have two files named `streamlit-exp-<something>`? This is your *source* code. You should note have experimental scripts here. Only put the streamlit files you plan to use for your frontend here. And give them better names. Nothing in source code should ever be an experiment!

File: src/visualization/streamlit-exp-v2.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    # generate list of labels and randomly insert into pandas dataframe
+    labels = np.random.choice(label_names, size=len(df))
+    df.insert(loc=0, column='label', value=labels)
+
+    return df
+
+
+@st.cache
+def gather_umap_data():
+    cwd = Path.cwd()
```
Comment (rkingery) [inline src/visualization/streamlit-exp-v2.py:36]: In your source code, always give your functions docstrings. They can start with simple one line descriptions of what the function does. Later on you'll want to expand them into sphinx format and make them prettier.

File: src/visualization/streamlit-exp-v2.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    filename1 = 'all_umap_pseudo_embeddings.json'
+    filename2 = 'centers_umap_pseudo_embeddings.json'
+    path_data = cwd/'data'/'current_centroid_data'
```
Comment (rkingery) [inline src/visualization/streamlit-exp-v2.py:39]: You sure you want to hard code these path names in like this here? If you end up changing them at all this function will break. Consider moving these names to the `config.py` and importing them in. This goes for *any* path or filename that shows up in your source code. Hardcoding these is dangerous, because you might delete them, or rename them, or move them, in which case things will break.

File: src/data/make_dataset.py [hunk 5]
```diff
@@ -1,30 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (rkingery) [inline src/data/make_dataset.py:1]: Why was this file removed?

File: requirements.txt (hunk 1 above)
Comment (earny-joe) [reply requirements.txt]: I just pushed a commit that updated `requirements.txt` to minimal libraries, while renaming the original file `requirements_experiments.txt`.

File: src/visualization/streamlit-exp-v2.py (hunk 2 above)
Comment (earny-joe) [reply src/visualization/streamlit-exp-v2.py]: In the most recent commit, I updated this to reflect a better naming convention. The v1 script (which was more experimentation than v1) was relocated to the `experiments` folder, while the v2 script was updated to `streamlit-v1.py` within the `src/visualization` folder.

File: src/visualization/streamlit-exp-v2.py (hunk 3 above)
Comment (earny-joe) [reply src/visualization/streamlit-exp-v2.py]: I added a docstring to the above. In my next branch, I'm expecting this to change, as this function's original purpose was to gather test data to ensure the streamlit/bokeh app produced a visualization.

File: src/visualization/streamlit-exp-v2.py (hunk 4 above)
Comment (earny-joe) [reply src/visualization/streamlit-exp-v2.py]: In my next branch, one of my first tasks will be to create a `config.py` file to avoid hardcoding these (amongst other things).

File: src/data/make_dataset.py (hunk 5 above)
Comment (earny-joe) [reply src/data/make_dataset.py:1]: I did not create it. I believe it was automatically generated when I initially created the repo.

File: requirements.txt [hunk 6]
```diff
@@ -1,249 +1,18 @@
-https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-2.2.5/en_core_web_sm-2.2.5.tar.gz
```
Comment (rkingery) [inline requirements.txt:1]: Did you make sure this new `requirements.txt` actually works first? Anytime you change this file you need to check that it actually works, as `pip install -r requirements.txt` is known to frequently break due to random package conflicts. To do so, create a temporary conda environment with minimal installs (only python and pip) and run `pip install -r requirements.txt`. If it works, good. Then try running some of your source scripts in this environment to ensure they work as well, and don't have weird issues with package versions coming up. As I said, anytime you change `requirements.txt` you should do this, because dependencies are very important to get right, and very finicky.

File: src/data/json_to_parq.py [hunk 7]
```diff
@@ -48,15 +48,15 @@ def clean_for_parquet(df):
 
 def parquet_to_gcs(df, day):
```
Comment (rkingery) [inline src/data/json_to_parq.py:49]: Add docstrings to these functions too.

File: src/data/json_to_parq.py [hunk 8]
```diff
@@ -48,15 +48,15 @@ def clean_for_parquet(df):
...
     df.to_parquet(
-        f'gs://thepanacealab_covid19twitter/dailies/'
+        'gs://thepanacealab_covid19twitter/dailies/'
         + f'{day}/{day}_tweets.parquet'
```
Comment (rkingery) [inline src/data/json_to_parq.py:52]: This is weird formatting. Consider putting this on one line and making the whole thing a format string. If flake8 yells at you, you should change the line character limit to 100 or higher, as I mentioned (default is like 70).

File: src/data/json_to_parq.py [hunk 9]
```diff
@@ -48,15 +48,15 @@ def clean_for_parquet(df):
...
     )
     print('Dataframe uploaded to bucket as parquet file.')
 
 
 def data_prep_wrapper(day):
     bucket_path = (
-        f'gs://thepanacealab_covid19twitter/dailies/'
+        'gs://thepanacealab_covid19twitter/dailies/'
```
Comment (rkingery) [inline src/data/json_to_parq.py:59]: Same as previous comment.

File: src/models/gen_embeddings.ipynb [hunk 10]
```diff
@@ -0,0 +1,722 @@
+{
```
Comment (rkingery) [inline src/models/gen_embeddings.ipynb:1]: Keep notebooks out of your src files. If something essential for the source code is currently in a notebook, convert it to a script and put that here. You want all your source code, even stuff that only runs periodically, to be scripts. 

There's *a lot* of stigma in the software world against people using notebooks for shared code and production code, and you doing it could make you be perceived the wrong way. This may change in the future, but it hasn't yet for sure. Keep notebooks just for experiments and personal stuff. Until it becomes the accepted norm, try to avoid using them for anything else.

File: src/visualization/streamlit-v1.py [hunk 11]
```diff
@@ -33,6 +33,12 @@ def gen_random_data():
 
 @st.cache
 def gather_umap_data():
+    '''
```
Comment (rkingery) [inline src/visualization/streamlit-v1.py:36]: These line character lengths are tiny. Update your flake8 config so you don't have to deal with this anymore.

File: .travis.yml [hunk 12]
```diff
@@ -6,4 +6,4 @@ install:
 # suppress flag is temporary, will be removed when tests are written
 script:
  - pytest --suppress-no-test-exit-code
```
Comment (rkingery) [inline .travis.yml:8]: I'm worried about this being in your configs and causing tests to pass that shouldn't be once you start writing them. Go ahead and remove the `--suppress-no-test-exit-code` flag, and instead write a dummy test in your `/tests` directory. Call it `test_dummy.py`, and add a single function `test_dummy` that has a single line `assert 1 == 1`. Then run `pytest` in your repo root to verify this test is running and passing. You'll add more very soon...

File: requirements.txt (hunk 6 above)
Comment (earny-joe) [reply requirements.txt:1]: Checked new `requirements.txt` with new conda environment. Ran `pip install -r requirements.txt` and there are no issues with the install. I ran tests (with new conda environment) on scripts in `src/data` and `src/visualization` and they worked.

File: src/data/json_to_parq.py (hunk 7 above)
Comment (earny-joe) [reply src/data/json_to_parq.py:49]: Added docstrings to all functions in `json_to_parq.py` file.

File: src/data/json_to_parq.py (hunk 8 above)
Comment (earny-joe) [reply src/data/json_to_parq.py:52]: Added `.flake8` config file to repo root, that increases the line limit to 100 (and ignores files within `experiment` folder).

File: .travis.yml (hunk 12 above)
Comment (earny-joe) [reply .travis.yml:8]: Added `test_dummy.py` with `test_dummy` function. Ran `pytest` command and "test" successfully ran.

**Comment by rkingery** [comment]
Two more things:

- Remove all `.py` files in your repo that are currently empty. Looks like cookiecutter created a lot of them. But you don't want them there.
- Squash your commits. You should always do this before merging a PR. It collapses all your commits for a branch into a single commit, that way it's easier to group commits together and figure out what was done where. See here on how to do this: https://gist.github.com/patik/b8a9dc5cd356f9f6f980
